### ISSUE TITLE
Usage of Objects (OOP) in R

### ISSUE THREAD

**Issue by msevestre** [Open-Systems-Pharmacology/OSPSuite-R#53]
http://adv-r.had.co.nz/OO-essentials.html

**Comment by msevestre** [comment]
We need to decide what to use, if at all

**Comment by ju-rgen** [comment]
I found this one a good introduction to OO in R: 
https://www.datamentor.io/r-programming/object-class-introduction/ 
https://www.datamentor.io/r-programming/reference-class/ 
https://www.datamentor.io/r-programming/inheritance/ 
etc. 
  
This also seems to be ok, but seems to be focused more on S3,S4 – probably nothing new compared to the datamentor pages: 
http://adv-r.had.co.nz/OO-essentials.html (as suggested from msevestre)
http://adv-r.had.co.nz/R5.html 
etc. 
  
Some more details on programming reference classes you find here: 
https://www.rdocumentation.org/packages/methods/versions/3.5.0/topics/ReferenceClasses

**Comment by ju-rgen** [comment]
From my perspective the usage of "tool classes" supports structuring the code.
The usage of "material/data classes" instead of named lists has the advantage, 
that properties are statically defined and "typed" somehow and cannot be misspelled or used with different types by chance.

**Comment by msevestre** [comment]
I think we have to be very careful here and understand WHO our user base is going to be.
I do not believe they will be software engineers and understand what OOP means. Therefore, usage of RC etc..should only be used internally. Understanding how RC works (aside from knowing how to call a method using the $ symbol and understanding the concept of side effects) should not be required.

Functional programming has the benefit of not having side effects, which makes the code much easier to reason about. 

In short, my suggestion would be to try to keep the class hierarchy as simple as possible, if at all,  and not force the user to define their own class to be able to use our system.  I am still not convinced that we need an object system for this library. The only requirements that I see at the moment is:
1. being able to define structure and extend them
1. pass function reference as parameter.  (to override default behavior)

I am not sure if this is possible in R but this is about it

**Comment by ju-rgen** [comment]
I agree: user should not need to know more about classes than the mentioned facts. 
And for interfaces the definition of structure is important. Here in my opinion the usage of named lists is not sufficient.
The creation of an object/instance for a class and assignment of values to its typed properties seems to be easy enough, that users can do this.

How intensive OOP is used in the internal implementation of the library, is a decision of the implementation team. (With regard to this it is just my private opinion, that I like structuring code via OOP - but I am not member of that team)

**Comment by DiedrichC** [comment]
I fully agree with Michael. We should be careful to not over-engineer the code. Note that "user" in the context of the SPM reporting engine will usually be someone who will have to adapt existing code on a pretty regular basis.

**Comment by ju-rgen** [comment]
See [here ](https://github.com/Open-Systems-Pharmacology/TFL-Library/issues/25)for example of usage of ReferenceClasses with fields.

**Comment by msevestre** [comment]
We will use ReferenceClass in our implementations
@PavelBal @abdullahhamadeh @pchelle: Please focus your reading especially on

http://adv-r.had.co.nz/R5.html

https://www.datamentor.io/r-programming/reference-class/


http://charlotte-ngs.github.io/RExperimentsWithS4RC/ExperimentsWithS4AndRef.html

**Comment by msevestre** [comment]
All: We'll try in the next few weeks using the S6 package for class. Very similar to RC but with some nicer syntax overall
https://adv-r.hadley.nz/r6.html

**Comment by PavelBal** [comment]
Close?

**Comment by msevestre** [comment]
Yep
